src/wasm/wasm-type.cpp

```diff
@@ -45,300 +48,284 @@ struct TypeInfo {
-    {TypeInfo({Type::i31ref}), Type::i31ref},
-    {TypeInfo(HeapType(HeapType::I31Kind), false), Type::i31ref},
-  };
+  // Track unique_ptrs for constructed types to avoid leaks.
+  std::vector<std::unique_ptr<Info>> constructedTypes;
 
-  uintptr_t canonicalize(const TypeInfo& info) {
+  // Maps from constructed types to their canonical Type IDs.
+  std::unordered_map<Info, uintptr_t> typeIDs;
+
+  typename Info::type_t canonicalize(Info info) {
     std::lock_guard<std::mutex> lock(mutex);
-    auto indexIt = indices.find(info);
-    if (indexIt != indices.end()) {
-      return indexIt->second;
+    auto indexIt = typeIDs.find(info);
+    if (indexIt != typeIDs.end()) {
+      return typename Info::type_t(indexIt->second);
     }
-    auto ptr = std::make_unique<TypeInfo>(info);
+    auto ptr = std::make_unique<Info>(info);
     auto id = uintptr_t(ptr.get());
     constructedTypes.push_back(std::move(ptr));
-    assert(id > Type::_last_basic_id);
-    indices[info] = id;
-    return id;
+    assert(id > Info::type_t::_last_basic_type);
+    typeIDs[info] = id;
+    return typename Info::type_t(id);
   }
+};
 
-  Type makeType(const Tuple& tuple) {
-    auto& types = tuple.types;
-#ifndef NDEBUG
-    for (Type t : types) {
-      assert(t.isSingle());
-    }
-#endif
-    if (types.size() == 0) {
-      return Type::none;
-    }
-    if (types.size() == 1) {
-      return types[0];
+struct TypeStore : Store<TypeInfo> {
+  Type canonicalize(TypeInfo info) {
+    if (info.isTuple()) {
+      if (info.tuple.types.size() == 0) {
+        return Type::none;
+      }
+      if (info.tuple.types.size() == 1) {
+        return info.tuple.types[0];
+      }
     }
-    return Type(canonicalize(TypeInfo(tuple)));
-  }
-
-  Type makeType(const HeapType& heapType, bool nullable) {
-#ifndef NDEBUG
-    switch (heapType.kind) {
-      case HeapType::FuncKind:
-      case HeapType::ExternKind:
-      case HeapType::AnyKind:
-      case HeapType::EqKind:
-      case HeapType::I31Kind:
-      case HeapType::ExnKind:
-      case HeapType::SignatureKind:
-        break;
-      case HeapType::StructKind:
-        for (Field f : heapType.struct_.fields) {
-          assert(f.type.isSingle());
+    if (info.isRef() && info.ref.heapType.isBasic()) {
+      if (info.ref.nullable) {
+        switch (info.ref.heapType.getBasic()) {
+          case HeapType::func:
+            return Type::funcref;
+          case HeapType::ext:
+            return Type::externref;
+          case HeapType::exn:
+            return Type::exnref;
+          case HeapType::any:
+            return Type::anyref;
+          case HeapType::eq:
+            return Type::eqref;
+          case HeapType::i31:
+            break;
         }
-        break;
-      case HeapType::ArrayKind:
-        assert(heapType.array.element.type.isSingle());
-        break;
+      } else {
+        if (info.ref.heapType == HeapType::i31) {
+          return Type::i31ref;
```

Comment:
Any reason to change this from map lookup to switch-case jump table? Wouldn't map lookup be cheaper?

Reply:
I'm not sure about the performance - the map lookup at least requires taking a lock and calculating a hash. I wouldn't expect this to have a measurable impact (but I haven't measured it). The reason I changed it was that what we had before was kind of a hack. It used tuple `TypeInfo`s to store individual `BasicType`s even though empty and singleton tuples were otherwise not allowed anywhere. In contrast, there is no need for `HeapTypeInfo` to be able to represent a `BasicHeapType` and it seemed cleaner to make that separation of concerns explicit for `TypeInfo` as well.

Reply:
Hmm, I thought it would be simpler if every `Type` has its matching `TypeInfo` and we can get one using the other, but as long as it is implementation detail and well encapsulated from outside I think either way should be fine.

Reply:
Thanks :) This change should not be observable outside of this function.